modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java

```diff
@@ -0,0 +1,149 @@
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpRequest;
+
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ContextPreservingActionListener;
+import org.elasticsearch.common.TriConsumer;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.http.BasicHttpRequest;
+import org.elasticsearch.rest.RestRequest;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+
+public class HttpHeadersAuthenticator {
+
+    private final TriConsumer<BasicHttpRequest, Channel, ThreadContext> populateThreadContext;
+    private final BiConsumer<BasicHttpRequest, ActionListener<Void>> authenticate;
+    private final ThreadContext threadContext;
+
+    public HttpHeadersAuthenticator(
+        TriConsumer<BasicHttpRequest, Channel, ThreadContext> populateThreadContext,
+        BiConsumer<BasicHttpRequest, ActionListener<Void>> authenticate,
+        ThreadContext threadContext
+    ) {
+        this.populateThreadContext = populateThreadContext;
+        this.authenticate = authenticate;
+        this.threadContext = threadContext;
+    }
+
+    public static final HttpHeadersAuthenticator NOOP = new HttpHeadersAuthenticator(null, null, null) {
+        @Override
+        public DefaultHttpRequest wrapNewMessage(DefaultHttpRequest decodedNewMessage) {
+            return decodedNewMessage;
+        }
+
+        @Override
+        public void authenticateMessage(HttpRequest request, Channel channel, ActionListener<Void> listener) {
+            listener.onResponse(null);
+        }
+    };
+
+    public DefaultHttpRequest wrapNewMessage(DefaultHttpRequest decodedNewMessage) {
+        HttpHeadersWithAuthenticationContext httpHeadersWithAuthenticationContext = new HttpHeadersWithAuthenticationContext(
+            decodedNewMessage.headers()
+        );
+        return new DefaultHttpRequest(
+            decodedNewMessage.protocolVersion(),
+            decodedNewMessage.method(),
+            decodedNewMessage.uri(),
+            httpHeadersWithAuthenticationContext
+        );
+    }
+
+    public void authenticateMessage(HttpRequest request, Channel channel, ActionListener<Void> listener) {
+        assert request.headers() instanceof HttpHeadersAuthenticator.HttpHeadersWithAuthenticationContext;
+        final BasicHttpRequest requestToAuthenticate = wrapToBasicHttpRequest(request);
+        final Supplier<ThreadContext.StoredContext> emptyContext = threadContext.wrapRestorable(threadContext.newStoredContext());
+        final ActionListener<Void> contextPreservingListener = new ContextPreservingActionListener<>(emptyContext, listener);
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            populateThreadContext.apply(requestToAuthenticate, channel, threadContext);
```

Comment:
Populate the thread context before authentication, exactly the same way it is currently populated for "REST" requests.